File: Lista01/GO/ex14.go
package main

import (
	"fmt"
	"sync"
)

// Definindo a struct Configuracao (a classe Singleton)
type Configuracao struct {
	Nome string
}

var instancia *Configuracao
var once sync.Once // Usado para garantir que a instância seja criada apenas uma vez

// Função para obter a instância da configuração (Singleton)
func GetInstancia(nome string) *Configuracao {
	once.Do(func() { // Executa apenas uma vez
		instancia = &Configuracao{Nome: nome}
	})
	return instancia
}

func main() {
	// Tentando criar duas instâncias
	config1 := GetInstancia("Configuração 1")
	config2 := GetInstancia("Configuração 2")

	// Exibindo os nomes das instâncias
	fmt.Println(config1.Nome) // Saída: Configuração 1
	fmt.Println(config2.Nome) // Saída: Configuração 1 (mesma instância)

	// Verificando se ambas apontam para a mesma instância
	if config1 == config2 {
		fmt.Println("Ambas as variáveis apontam para a mesma instância.")
	} else {
		fmt.Println("As variáveis apontam para instâncias diferentes.")
	}
}

File: Lista01/Java/ex14.java
public class ex14 {
    // Atributo estático para armazenar a instância única
    private static ex14 instancia;

    // Atributos de configuração (exemplo)
    private String nomeBanco;
    private String urlBanco;

    // Construtor privado para evitar a instância externa
    private ex14() {
        // Inicializações de configuração
        nomeBanco = "MeuBanco";
        urlBanco = "jdbc:meubanco://localhost:3306/minha_base";
    }

    // Método estático para obter a instância única
    public static ex14 getInstancia() {
        if (instancia == null) {
            instancia = new ex14();
        }
        return instancia;
    }

    // Métodos de configuração (getters)
    public String getNomeBanco() {
        return nomeBanco;
    }

    public String getUrlBanco() {
        return urlBanco;
    }

    // Método para testar a classe Singleton
    public static void main(String[] args) {
        ex14 config1 = ex14.getInstancia();
        ex14 config2 = ex14.getInstancia();

        // Verificando se ambas referências apontam para a mesma instância
        System.out.println("Configuração 1: " + config1.getNomeBanco());
        System.out.println("Configuração 2: " + config2.getNomeBanco());
        System.out.println("As instâncias são iguais? " + (config1 == config2)); // true
    }
}
